File: spelling-bee/tools/generate-puzzles.py
"""Builds the puzzle set embedded in index.html.

Two lists per puzzle, because no single "is this word common enough" cutoff works:
too strict and the game rejects words everyone knows (appall, attune, adduce);
too loose and answer lists hit 120 words and the top rank becomes unreachable.

  words  -- REQUIRED. Sets the maximum score, the ranks, the "N to find" count,
            and the top rank. Common vocabulary only.
  bonus  -- ACCEPTED but never required. Real words that are rarer (crouton,
            beignet, percipient). They score, they just aren't needed for 100%.

Commonness comes from wordfreq's Zipf scale over a real corpus, which is the only
signal that reliably separates rare-but-real (percipient, 1.17) from dictionary
padding (alaap, apoop, eevn -- all exactly 0.00).

    pip install wordfreq
    npm install word-list
    python3 tools/generate-puzzles.py
"""
import json, os, re, sys
from collections import defaultdict

try:
    from wordfreq import zipf_frequency
except ImportError:
    sys.exit("pip install wordfreq")

HERE = os.path.dirname(os.path.abspath(__file__))
DICT = os.path.join(HERE, 'node_modules', 'word-list', 'words.txt')
if not os.path.exists(DICT):
    DICT = os.path.join(HERE, '..', '..', 'node_modules', 'word-list', 'words.txt')
if not os.path.exists(DICT):
    sys.exit("npm install word-list (looked in tools/ and the repo root)")

REQUIRED_MIN = 2.5   # common enough that the top rank stays fair
BONUS_MIN    = 1.0   # real word, just rarer; below this is dictionary padding
N_PUZZLES    = 14

# Slurs, explicit vulgarities and informal non-words. Screened against the full
# pool, not guessed. Ordinary words with ordinary meanings are deliberately NOT
# here (tart, cock, cripple, queer, moron, oriental) -- rejecting those would
# just tell an honest player "not in word list".
BLOCKED = set("""
kike coon chink dago gook honky hymie injun kaffir mick paddy polack quadroon
negro negroid wetback wop yid spic pickaninny darkie darky golliwog raghead
towelhead jewed gypped gyp faggot dyke poofter tranny catamite
retard retarded cretin imbecile mongoloid midget harelip halfwit gimp spastic spaz
fuck fucked fucker cunt twat wank wanker nookie whore whored floozy
outta gonna wanna gotta kinda lotta oughta dunno cuz nuthin
""".split())

# ---------------------------------------------------------------- word pools
raw = [w.strip() for w in open(DICT, encoding='utf-8')]
pool, freq = [], {}
for w in raw:
    if not re.fullmatch(r'[a-z]{4,}', w) or 's' in w or w in BLOCKED:
        continue
    f = zipf_frequency(w, 'en')
    if f >= BONUS_MIN:
        pool.append(w)
        freq[w] = f
required_pool = {w for w in pool if freq[w] >= REQUIRED_MIN}
print(f"pool {len(pool)} words (zipf>={BONUS_MIN}), of which {len(required_pool)} "
      f"are common enough to require (zipf>={REQUIRED_MIN})")

mask = lambda w: sum(1 << (ord(c) - 97) for c in set(w))
allwords = set(raw)

# ------------------------------------------------- morphology (derived forms)
def bases(w):
    out, n = [], len(w)
    for suf, k in (('ing', 3), ('ed', 2), ('er', 2), ('est', 3)):
        if w.endswith(suf) and n > k + 2:
            b = w[:-k]
            out += [b, b + 'e', b[:-1] if len(b) > 2 and b[-1] == b[-2] else None, b + 'y']
    if w.endswith('ly') and n > 4: out += [w[:-2], w[:-2] + 'y', w[:-2] + 'le']
    if w.endswith('y') and n > 4:  out += [w[:-1], w[:-1] + 'e']
    for suf in ('ment', 'ness', 'able', 'ible', 'ful', 'less', 'ish', 'ion', 'ity'):
        if w.endswith(suf) and n > len(suf) + 3: out += [w[:-len(suf)], w[:-len(suf)] + 'e']
    return [b for b in out if b]

_dcache = {}
def derived(w):
    if w not in _dcache:
        _dcache[w] = any(b in allwords for b in bases(w))
    return _dcache[w]

# ------------------------------------------------------------ index by letters
by_mask = defaultdict(list)
for w in pool:
    by_mask[mask(w)].append(w)

def score(w, sm):
    s = 1 if len(w) == 4 else len(w)
    if len(w) >= 7: s += 3
    if mask(w) == sm: s += 14 if len(w) == 7 else 7
    return s

# candidate letter sets: any COMMON word with exactly 7 distinct letters, so
# every puzzle is guaranteed a pangram people have actually heard of
sets_ = {mask(w) for w in required_pool if len(set(w)) == 7}
print(f"candidate letter sets: {len(sets_)}")

cands = []
for sm in sets_:
    bucket, sub = [], sm
    while True:
        bucket += by_mask.get(sub, [])
        if sub == 0: break
        sub = (sub - 1) & sm
    for bit in range(26):
        if not sm >> bit & 1: continue
        c = chr(97 + bit)
        hit = [w for w in bucket if c in w]
        req = sorted(w for w in hit if w in required_pool)
        bon = sorted(w for w in hit if w not in required_pool)
        if not 25 <= len(req) <= 60: continue
        pans = [w for w in req if mask(w) == sm]
        if len(pans) < 2: continue
        long_ = [w for w in req if len(w) >= 7]
        if len(long_) / len(req) < 0.25: continue
        if max(len(w) for w in req) < 9: continue
        dfrac = sum(derived(w) for w in req) / len(req)
        if dfrac > 0.42: continue
        rootlong = [w for w in long_ if not derived(w)]
        if len(rootlong) < 6: continue
        cands.append(dict(
            sm=sm, center=c,
            outer=[chr(97 + i) for i in range(26) if sm >> i & 1 and i != bit],
            words=req, bonus=bon, pangrams=pans,
            perfect=any(len(p) == 7 for p in pans),
            dfrac=dfrac, rootlong=len(rootlong),
            max=sum(score(w, sm) for w in req),
            maxlen=max(len(w) for w in req),
            ing=all(sm >> b & 1 for b in (8, 13, 6)),
        ))
print(f"qualifying puzzles: {len(cands)}")

quality = lambda p: (min(len(p['pangrams']), 4) * 9 + (10 if p['perfect'] else 0)
                     + p['rootlong'] * 2.2 + min(p['maxlen'], 12) * 0.8
                     + (1 - p['dfrac']) * 25 - abs(len(p['words']) - 38) * 0.3)
cands.sort(key=quality, reverse=True)

from collections import Counter
chosen, centers, ings = [], Counter(), 0
popcount = lambda m: bin(m).count('1')
for p in cands:
    if len(chosen) >= N_PUZZLES: break
    if any(popcount(c['sm'] & p['sm']) >= 5 for c in chosen): continue
    if centers[p['center']] >= 2: continue
    if p['ing'] and ings >= 3: continue
    if p['ing']: ings += 1
    centers[p['center']] += 1
    chosen.append(p)

for i, p in enumerate(chosen, 1):
    print(f"{i:2}. {p['center'].upper()} | {''.join(p['outer']).upper()}  "
          f"{len(p['words']):2} required + {len(p['bonus']):3} bonus, "
          f"{len(p['pangrams'])} pangram(s){'*' if p['perfect'] else ''}, {p['max']} pts")
    print(f"    pangrams: {', '.join(p['pangrams'])}")

out = [dict(id=f"hc-{i:03d}", center=p['center'], outer=p['outer'],
            pangrams=p['pangrams'], words=p['words'], bonus=p['bonus'])
       for i, p in enumerate(chosen, 1)]
dest = os.path.join(HERE, 'puzzles.json')
json.dump(out, open(dest, 'w'), indent=2)
print(f"\nwrote {dest}")

File: spelling-bee/tools/generate-puzzles.mjs
import { createRequire } from 'module';
import fs from 'fs';
const require = createRequire(import.meta.url);

const TIERS = [10, 20, 35, 40, 50, 55];
const DIALECTS = ['english', 'american'];

// ---- words excluded from every answer list ------------------------------
// Ethnic / religious / disability / sexual-orientation slurs, explicit
// vulgarities, and informal contractions that aren't really words. Ordinary
// words with ordinary meanings (tart, cock, cripple, queer) are NOT blocked --
// excluding those would just surprise players with "not in word list".
const BLOCKED = new Set([
  'kike', 'coon', 'chink', 'dago', 'gook', 'honky', 'mick', 'wetback',
  'negro', 'negroid', 'faggot', 'dyke', 'poofter',
  'retard', 'retarded', 'cretin', 'midget', 'harelip', 'halfwit',
  'fuck', 'cunt', 'twat', 'wank', 'nookie', 'whore', 'turd',
  'outta', 'gonna', 'wanna', 'gotta', 'kinda', 'lotta', 'oughta',
]);

const raw = new Set();
for (const t of TIERS) for (const d of DIALECTS) {
  let list; try { list = require(`wordlist-english/${d}-words-${t}.json`); } catch { continue; }
  for (const w of list) raw.add(w);
}
const allWords = new Set([...raw].filter(w => /^[a-z]+$/.test(w)));
const pool = [...allWords]
  .filter(w => w.length >= 4 && !w.includes('s') && !BLOCKED.has(w))
  .sort();



const maskOf = w => { let m = 0; for (const c of w) m |= 1 << (c.charCodeAt(0) - 97); return m; };
const popcount = m => { let n = 0; while (m) { m &= m - 1; n++; } return n; };

// ---- morphology: is this word just a shorter word plus a routine ending? ----
const bases = (w) => {
  const out = [];
  const strip = (n) => w.slice(0, w.length - n);
  const withE = (n) => strip(n) + 'e';
  const undouble = (n) => {
    const b = strip(n);
    return b.length > 2 && b[b.length - 1] === b[b.length - 2] ? b.slice(0, -1) : null;
  };
  const unY = (n) => strip(n) + 'y';
  for (const [suf, n] of [['ing', 3], ['ed', 2], ['er', 2], ['est', 3]]) {
    if (w.endsWith(suf) && w.length > n + 2) {
      out.push(strip(n), withE(n), undouble(n), unY(n));
    }
  }
  if (w.endsWith('ly') && w.length > 4) out.push(strip(2), unY(2), strip(2) + 'le');
  if (w.endsWith('y') && w.length > 4) out.push(strip(1), withE(1), undouble(1));
  for (const suf of ['ment', 'ness', 'able', 'ible', 'ful', 'less', 'ish', 'ion', 'ity']) {
    if (w.endsWith(suf) && w.length > suf.length + 3) out.push(strip(suf.length), strip(suf.length) + 'e');
  }
  return out.filter(Boolean);
};
const derivedCache = new Map();
const isDerived = (w) => {
  if (!derivedCache.has(w)) derivedCache.set(w, bases(w).some(b => allWords.has(b)));
  return derivedCache.get(w);
};

const byMask = new Map();
for (const w of pool) {
  const m = maskOf(w);
  if (!byMask.has(m)) byMask.set(m, []);
  byMask.get(m).push(w);
}
console.log(`pool: ${pool.length} words | derived (inflected/suffixed): ${pool.filter(isDerived).length}`);

const sets = new Set();
for (const [m] of byMask) if (popcount(m) === 7) sets.add(m);

const scoreWord = (w, setMask) => {
  const pangram = maskOf(w) === setMask;
  let s = w.length === 4 ? 1 : w.length;
  if (w.length >= 7) s += 3;
  if (pangram) s += (w.length === 7 ? 14 : 7);
  return s;
};

const candidates = [];
for (const setMask of sets) {
  const words = [];
  for (let sub = setMask; ; sub = (sub - 1) & setMask) {
    const b = byMask.get(sub); if (b) words.push(...b);
    if (sub === 0) break;
  }
  for (let bit = 0; bit < 26; bit++) {
    if (!(setMask & (1 << bit))) continue;
    const center = String.fromCharCode(97 + bit);
    const ws = words.filter(w => w.includes(center)).sort();
    if (ws.length < 25 || ws.length > 55) continue;
    const pangrams = ws.filter(w => maskOf(w) === setMask);
    if (pangrams.length < 2) continue;                       // one pangram makes the top rank a coin flip
    const derived = ws.filter(isDerived);
    const derivedFrac = derived.length / ws.length;
    if (derivedFrac > 0.42) continue;                        // reject gerund/inflection padding
    const rootLong = ws.filter(w => w.length >= 7 && !isDerived(w));
    if (rootLong.length < 6) continue;                       // needs real long-word depth
    const maxLen = Math.max(...ws.map(w => w.length));
    if (maxLen < 9) continue;
    if (ws.filter(w => w.length >= 7).length / ws.length < 0.25) continue;   // spec §6.4: 25% long
    const perfect = pangrams.some(p => p.length === 7);
    candidates.push({
      setMask, center,
      outer: [...Array(26).keys()].filter(i => (setMask & (1 << i)) && i !== bit).map(i => String.fromCharCode(97 + i)),
      words: ws, pangrams, perfect, derivedFrac,
      rootLong: rootLong.length,
      max: ws.reduce((a, w) => a + scoreWord(w, setMask), 0),
      maxLen,
      ing: (setMask & (1 << 8)) && (setMask & (1 << 13)) && (setMask & (1 << 6)),  // i,n,g all present
    });
  }
}
console.log(`qualifying puzzles: ${candidates.length}`);

const quality = p =>
  Math.min(p.pangrams.length, 4) * 9 +
  (p.perfect ? 10 : 0) +
  p.rootLong * 2.2 +
  Math.min(p.maxLen, 12) * 0.8 +
  (1 - p.derivedFrac) * 25 +
  -Math.abs(p.words.length - 38) * 0.3;
candidates.sort((a, b) => quality(b) - quality(a));

const chosen = [];
const centerUse = new Map();
let ingCount = 0;
for (const p of candidates) {
  if (chosen.length >= 14) break;
  if (chosen.some(c => popcount(c.setMask & p.setMask) >= 5)) continue;   // sets must differ by 3+ letters
  if ((centerUse.get(p.center) || 0) >= 1) continue;                      // every puzzle a different center
  if (p.ing && ingCount >= 3) continue;                                   // cap the -ING-friendly sets
  if (p.ing) ingCount++;
  centerUse.set(p.center, 1);
  chosen.push(p);
}

console.log(`\nselected ${chosen.length} puzzles:\n`);
for (const [i, p] of chosen.entries()) {
  console.log(`${String(i + 1).padStart(2)}. ${p.center.toUpperCase()} | ${p.outer.join(' ').toUpperCase()}  ` +
    `${String(p.words.length).padStart(2)}w, ${p.pangrams.length} pan${p.perfect ? '*' : ' '}, ` +
    `${p.rootLong} root-long, ${Math.round(p.derivedFrac * 100)}% derived, ${p.max} pts`);
  console.log(`    pangrams: ${p.pangrams.join(', ')}`);
  console.log(`    longest:  ${[...p.words].sort((a, b) => b.length - a.length).slice(0, 6).join(', ')}`);
}

fs.writeFileSync(new URL('./puzzles.json', import.meta.url), JSON.stringify(
  chosen.map((p, i) => ({
    id: `hc-${String(i + 1).padStart(3, '0')}`,
    center: p.center, outer: p.outer, pangrams: p.pangrams, words: p.words,
  })), null, 2));
console.log('\nwrote puzzles.json');
